### ISSUE TITLE
[feature request] possibility to detect if a key is pressed on keyboard or joystick

### ISSUE THREAD

**Issue by amaendle** [kollokollo/X11Basic#11]
I was trying to translate an old GFA-basic jump&run to X11Basic, but I cannot make the controls work.

Is there no function or workaround to detect if a key is _currently_ pressed on the keyboard?

KEYEVENT does not help, because I cannot detect if or when the key is being released.

INKEY$ does not detect anything if the graphic window is active.

And concerning the joystick: Stick() and Strig() are not implemented.

**Comment by kollokollo** [comment]
INKEY$ will not work (only in the terminal), but what about MOUSEEVENT ? You could use the muse buttons (and shift key, ctrl key, alt key). And: Have you tried the EVENT command? It has a lot of parameters, but it should be able to detect key releases for all keys on the keyboard. There is an example program which demontrates howi works: https://codeberg.org/kollo/X11-Basic_examples/src/branch/master/All/Tests/eventtest.bas

**Comment by amaendle** [comment]
The EVENT command sounds good. I will try it out and let you know.

**Comment by amaendle** [comment]
I had a look at the eventtest.bas. The events on the keyboard are recognized as mouse events (e.g. pressing the key produces a mouse button press event). This is strange, but not really a problem.

However, it still will not work for me in this way. Assume the loop in eventtest.bas is running: When I press a key, then `event?(4)=true`. Then `EVENT t,x,y,xroot,yroot,s,k,ks,t$,tim` is called and **waits** for the next  event **until another event happens**,  e.g. I release the key. The parameters of the event which was pending according to `EVENT?` (i.e. before `EVENT`) cannot be accessed. So I cannot know if a key has been pressed and not released. Furthermore `EVENT`  is inappropriate, because it stops the loop until a new event happens. It should be somehow possible to pick events from a stack an dispatch them after accessing their information...

What would help me is something like EVENT?, which not only tells me that a key has been pressed (or released), but also which key.

**Comment by kollokollo** [comment]
Yes, you are right, the EVENT functions always block. It would be difficult to workaround. Either use separate threads (--> SPAWN) or implement an event queue. Both possible in X11-Basic but of course not convenient. BTW: The mouse button state can be polled with MOUSE x,y,k.

**Comment by amaendle** [comment]
Do you have any hints how to implement the event queue? I actually don't see how this can be done in X11-Basic. 

I suppose you would have to use SPAWN also for this?

**Comment by kollokollo** [comment]
Whenever possible in your program flow (main loop), check if there is an event pending (--> EVENT?) and then receive the events (EVENT) whatever it is and set the state of all buttons and key accordingly (in regular variables or arrays). Then in your program you simple evaluate the state of the keys by checking variables. 
... your code here
WHILE EVENT?
   EVENT a,b,c,f...
  if a=KEYpressed
   keystate(b)=1
else if a=keyreleased
  keystate(b)=0
endif
wend
.. continue
if keystate(upkey)=1
   ....
else
 ...
endif

**Comment by kollokollo** [comment]
If you are only interested in key states, no event queue is needed. If you need the information which key was pressed first (the order of the events) then you need a queue (FIFO-Buffer). This would be more complicated to implement. But I think you do not need it.

**Comment by amaendle** [comment]
The order matters, because it is important if a key has first been released and then pressed again (key is active) or vice versa.

Also according to my experiments what you describe above does not work: I tried checking for e.g. keyreleased. Then EVENT is called. But as long as I keep the key pressed and don't create other events (e.g. by moving the mouse) nothing happens. So if I keep the key pressed for 10 seconds then the queue is stuck for these 10 seconds. And as soon as I release the key I get the information of the keyreleased event. The keypressed event gets just swallowed and never appears.

**Comment by kollokollo** [comment]
Hm, the idea was, that you only call EVENT when you know that there is definitely an event there (EVENT?=true), which you will immedeately get, But maybe there is a bug? I have not used it for a long time. Can of course be. Do you have a small example code which produces this hang? So I could also experiment with it.

**Comment by amaendle** [comment]
At the moment I'm busy, but I hope to post an example this weekend.
If it comes out that the problem is really a bug (and not me) I'll just use SPAWN...

**Comment by kollokollo** [comment]
Another thing, you should use the `EVENT?` function with a mask of 3 if you want it to recognize key press and key release events. If you use 4, it will only detect mouse press events. If you want every event, use `EVENT?(0xffffff)`, or `EVENT?(-1)`. (see also the users manual, page 283).

Also a statement 
`IF EVENT?(-1)=TRUE `
should be avoided, because `TRUE` means -1. Here you do not want exact match for -1 but just "anything but 0". So better use

`IF EVENT?(-1)`

without the "=TRUE"

(sorry, I used this myself in the pseudo-code above).

**Comment by amaendle** [comment]
Acccording to the manual you are right, but what actually happens on my Windows 10 computer is different.
If I use the mask of 3 no key presses are recognized at all! However, if I use the mask of 4 both, mouse presses and keyboard presses are recognized.

This is exactly the strange behaviour I mentioned above:

> I had a look at the eventtest.bas. The events on the keyboard are recognized as mouse events (e.g. pressing the key produces a mouse button press event). This is strange, but not really a problem.

**Comment by amaendle** [comment]
Ok, now one more try to explain, what is going wrong.  Actually the eventtest.bas should be a good demonstration, of what is wrong. However, I have simplified the code: [https://file.io/ZFa4h70foyu0](url)

The first problem of the original eventtest.bas: It does not even run well, but stops as soon as I press a key: There is a dimension error, because the limits of mb() are exceeded. This is because EVENT returns for key presses and releases the event types 4 and 5 (instead of 2 and 3). Therefore I have increased the dimension of mb() in my code.

When you start the program (don't move the mouse, to avoid mouse move events!) you will permanently get outputs in the console ("NO EVENT WAITING"). This proofs that the loop is running. If you then press a key on the keyboard (and hold it), there are no more text outputs in the console. This indicates that the loop is currently an hold (because EVENT? was true and subsequently EVENT was called). **This is a major problem for my purpose, because the main loop should keep running.**  As soon as you release the key, you get the regular outputs again - and in the graphic window the gprint command shows that a mouse button release was recognized (instead of a key button release). **This is a bug, not heavy, but confusing.**

The key press event was never recognized. Apparently EVENT has not taken the event which was waiting in the event queue.. Instead it has only waited for a new event. **That behavior matches the description in the manual. However, this is also the reason why your suggestion for the event handling does not help.**

I hope it is now clearer. For now I have put my weekend project to port the GFA basic code to X11 on hold. Even a solution with spawn is not possible, because a call of EVENT in the spawned function also pauses the main thread. (**Another bug?**)

**Comment by kollokollo** [comment]
Oh, I should have asked this before, you are using WINDOWS. That might all explain the strange behaviour. On my linux system (with the X11 / X.org) all events are fine, at least the eventtest demo works as expected. 

Well yes, these are all bugs on the WINDOWS port. X11-Basic was never adapted for newer WINDOWS versions (later than WINDOWS XP). I am sorry. I do not use WINDOWS myself and rely on other contributors adapting X11-Basic for WINDOWS. But I thought that you were talking about the Android version of X11-Basic. This is the one most users have. 
... but let me have a look at your example...
Hm. the link https://file.io/ZFa4h70foyu0 seems not to work. I get a 404 Page not found Error

**Comment by amaendle** [comment]
Sorry for the outdated link. This should work longer: https://pastebin.com/petft70Q

What a pity that the Windows version isn't maintained any more. X11 Basic seemed like a good pick to get an old GFA-basic game running on modern systems including Windows.

**Comment by kollokollo** [comment]
On linux, this program works (with a little slow reaction, but I think this comes from the PAUSE statements).

**Comment by amaendle** [comment]
I also tested it for fun on Android - it works ok. The only thing is that the keydown events are fired continuously until a keyup event occurs. I would have expected it to be fired only once per keypress.

I have no clue what is wrong in Windows. Do you have any idea where the error might be?

**Comment by kollokollo** [comment]
Yes on Android this is a little more tricky, because of the virtual on-screen-keyboard.You basically cannot hold down a key. There only a short or long tap can be produced. But on WINDOWS the SDL library is used and there this should be possible. However for a very long time I have not looked into that part of the code. It might be a simple error (like the one you found with MOUSEDOWN events Issue #13). When I find time, I can have a look. But it also might well be a problem with WINDOWS itself. BTW: Can you compile X11-Basic from sources for windows yourself? I can compile it but I cannot test it.

**Comment by amaendle** [comment]
At the moment I would happily test a compiled version. Compiling in Windows is such a mess and I currently I don't have a Linux machine for cross-compiling available.

BTW: For my Android tests I didn't use the virtual keyboard, but a USB keyboard.

**Comment by amaendle** [comment]
I tried to compile it in a Fedora VM. Both for the master and the unstable branch I get this error, after` make windows`:

functions.c:300:16: error: conflicting types for 'clog10'
  300 | static COMPLEX clog10(COMPLEX a) {return(complex_div(((COMPLEX (*)())clog)(a),FLOAT2COMPLEX(10)));}

Wich setting are you using to compile it? Is this because I'm using a 64 bit system?

**Comment by kollokollo** [comment]
Hm, I do not think so. I get the same, but just as a warning: 
`
functions.c:300:16: warning: conflicting types for built-in function ‘clog10’ [-                Wbuiltin-declaration-mismatch]                                                                  
 static COMPLEX clog10(COMPLEX a) {return(complex_div(((COMPLEX (*)())clog)(a),F                LOAT2COMPLEX(10)));}                                                                            
`

The command was: 
`i686-w64-mingw32-gcc -DWINDOWS -c xbasic.c tools.c loadprg.c file.c io.c io_basic.c parser.c variablen.c svariablen.c array.c parameter.c fft.c mathematics.c memory.c runtime.c wort_sep.c ltext.c functions.c sfunctions.c afunctions.c kommandos.c gkommandos.c bitmap.c do_gets.c errortxt.c window.c sysVstuff.c aes.c graphics.c sound.c number.c keywords.c decode.c virtual-machine.c vm-api.c type.c lodepng.c md5.c sha1.c WINDOWS/fnmatch.c -I./WINDOWS -I./WINDOWS/include -I./WINDOWS/include/GMP`

**Comment by amaendle** [comment]
with exactly the same command I still get the message as error

**Comment by kollokollo** [comment]
Yes, I have investigated: It is an issue with the new version of mingw. Try following: 

Remove the #define _GNU_SOURCE at the beginning of functions.c
(I need to find out/remember, why it is there), but it should solve that error.

**Comment by kollokollo** [comment]
OK, I found out why: The linux version now will not compile.

**Comment by kollokollo** [comment]
Here is a workaround: change functions.c

`#ifndef WINDOWS
#define _GNU_SOURCE
#endif
`
This should compile for both now. Does it work?

**Comment by kollokollo** [comment]
Hint: Instead of `make windows` you can just do a `make xbasic.exe`. Then only this .exe file will be rebuild. To build the compiler xbc.exe a full working installation of X11-Basic on the host-system (fedora) would be needed. I f you dont have this, you would get more errors.

**Comment by kollokollo** [comment]
Concerning the EVENT functions: I have looked into the code  (gkommandos.c):

Under WINDOWS (USE_SDL) the argument (mask) of `EVENT?()` is ignored. It becomes true at any event pending. 

And the command `EVENT` should return following events: SDL_QUIT, SDL_ACTIVEEVENT, SDL_KEYUP, SDL_KEYDOWN, SDL_MOUSEWHEEL, SDL_MOUSEBUTTONUP, SDL_MOUSEBUTTONDOWN, SDL_MOUSEMOTION, SDL_WINDOWEVENT, and also others which appear here.

**Comment by kollokollo** [comment]
Under WINDOWS (USE_SDL) the `KEYEVENT` command detects only SDL_KEYDOWN (and not  SDL_KEYUP) and will wait for such an event.

**Comment by amaendle** [comment]
yes, it compiles.

I realized that the mask is ignored and have also looked into parts of the SDL documentation... I compiled a simple test program in C using SDL_PollEvent(NULL) to check for events and then use SDL_WaitEvent(&event) if the event queue is not empty. It seemed to work. This makes me wonder what is going wrong in X11-Basic using event? and event.

Maybe some other functions are messing around with the events?

**Comment by kollokollo** [comment]
Hm, yes. Probably the function handle_window (in window.c) interfers. It is supposed to handle redraw/window exposure events.

**Comment by kollokollo** [comment]
It calls `handle_event`. This function removes events from the queue, but in fact does nothing with it.

**Comment by kollokollo** [comment]
The `EVENT?()` function can well detect such exposure/window events, which get removed before `EVENT` is called. And then EVENT will wait for the next event and block. To fix this, the `EVENT?()` function must analyze the pending events and then only return true on "real" events. 
Or: maybe one can call `graphics()` just before the `EVENT?()` function (graphics should have removed the window events by calling handle_window ). Maybe this can work. (would be easy to fix).

**Comment by kollokollo** [comment]
Yea, graphics(); actually is not called in the SDL part of `f_eventf()`.

**Comment by kollokollo** [comment]
Hm, maybe not. this would remove all events. I dont know if there is a way to put back the events...

**Comment by kollokollo** [comment]
Maybe the opposite: In `c_allevent` (gkommandos.c) `graphics()` is called.#elif defined USE_SDL

  SDL_Event event;
  char buf[4];
  graphics();
  if(SDL_WaitEvent(&event)==0) return;

Maybe graphics(); should be removed here... It is unlikely that `EVENT` is the first command in a basic program, so we can assume, that a window is already open.

**Comment by amaendle** [comment]
Might be a solution. But I wouldn't take it for granted that event is never the first command. Maybe graphics could have an optional argument `handlewindow` which is by default true. Within `c_allevent` the `graphics` function can be called with `handlewindow=FALSE`.

Also I stumbled upon `SDL_PeepEvents` [https://wiki.libsdl.org/SDL_PeepEvents](url) which permits to check only a range of SDL_EventTypes. Also you can check for events without deleting them from the queue.

Edit: okay, C doesn't have default arguments... but still I think the unfortunate case of EVENT at the beginning should also  be handled

**Comment by amaendle** [comment]
Anyway, I compiled it now after removing graphics(). It seems to work for me. Thanks a lot for putting so much effort into this. The keyboard issue might be solved.

Edit: It it works only for my little test program which only creates console output. My other program breaks now right from the start, because graphics() isn't called :/

Edit 2: I have to correct my previous statement. My program does not fail because of the missing graphics() statement. However it fails because of the command to read a 16bit Integer: `Read Infolang&`
It worked fine in the binary Windows version I downloaded. However, the version I compiled just returns "not supported". Why has this changed?

**Comment by kollokollo** [comment]
This might be another bug. Maybe better open a separate issue for it. 
Coming back to the EVENT problem. For now I am going to simply remove the graphics() statement. However, this is not perfect, because 
1. If `EVENT` is the first graphics command in the program, X11-Basic will crash
2. If you use another graphics command (like `PLOT`) between `EVENT?()` and `EVENT,` nothing is gained.

So in any case a real fix would involve filtering events (maybe with SDL_PeepEvents). I need to further investigate here. Maybe I can come up with a solution, but I would need someone to test it.

**Comment by kollokollo** [comment]
Aeh, BTW: X11-Basic does not have 16-bit integers. The suffix "&" is for arbitrary big integers. And this support needs to be compiled in (using libgmp). And it will be compiled into the WINDOWS version, but only, if the linux version which was used to crosscompile it, also has the gmp support compiled in. And for having this, on your fedora system you need to have libgmp installed and detected by the ./configure script. I know, it's complicated.

Edit: when you start X11-Basic (on fedora), you should see this line in the intro:
***  x11 la  rl  gmp fft     snd usb bt                ***
This shows you what compile you have:
x11 -- graphics = X.org (linux)
fb -- graphics = framebuffer (Android)
sdl -- graphics = SDL (Windows)
gem -- graphics = GEM (ATARI ST)
la -- linear algebra library in use
rl -- readline library in use (nice to have only on linux)
Pi -- support for wiring pi interface (Raspberry Pi)
gmp -- support for big integers
fft -- fast fourier transformation library 
snd -- sound support
usb -- usb support
bt -- bluetooth support
cry -- support for (strong) cryptography (forbidden in USA)
mqtt -- MQTT support

A normal WIndows compile should have:
***  sdl la      gmp         snd                       ***

**Comment by kollokollo** [comment]
I suggest that we close this issue and create new ones for the remaining problems. 

The compile error (mingw) should be fixed now and also the `EVENT?()` functions now should use the mask. The `graphics()` function  now will not remove all(!) events. So It should be possible now to detect, if there are pending MOUSE/KEYBOARD events. However, it needs to be tested if there is now a new problem, when the events (especialle MOUSEMOTION) are not taken. This could fill up the internal event queue and I have no Idea what happens, when this gets full. Maybe we will reopen this isue, or reference it and create new issues with upcoming problems.

**Comment by amaendle** [comment]
Fine for me. I am closing it.
